### DISCUSSION TITLE
Problem with instantiating `OAuth2AuthorizationCodeBearer` security scheme using injected `Settings` instance

### DISCUSSION THREAD

**Discussion by 15bubbles** [fastapi/fastapi#8709] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
Code examples are contained in description as it is probably easier to follow all the options I already tried
```

### Description

Hi guys!

First of all, thank you for all the good work on FastAPI!

Second of all, sorry if similar problem has been already posted but I somehow couldn't find anything related

Third of all, and getting straight to the point: my question revolves around settings, security Oauth2 class and dependency injection.

So we are using `OAuth2AuthorizationCodeBearer` as our security dependency. We have a function that uses this dependency and verifies if provided token is valid and if it contains required scopes. We then use this dependency in our path operation functions.

However to properly instantiate `OAuth2AuthorizationCodeBearer` we need some values from `Settings`.

So this makes me wondering how to instantiate this parametrized dependency using injected `Settings` instance and still have this properly return token and appear as security schema in OpenAPI docs and to render `Authorize` button in Swagger UI.

I already tried

```python
from typing import Any

from fastapi import Depends, Security
from fastapi.security import OAuth2AuthorizationCodeBearer, SecurityScopes

from our_project.settings import get_settings, Settings


def _prepare_scopes(scopes: str) -> dict[str, str]:
    ...


def get_oauth2_scheme(settings: Settings = Depends(get_settings)) -> OAuth2AuthorizationCodeBearer:
    return OAuth2AuthorizationCodeBearer(
        tokenUrl=settings.token_url,
        authorizationUrl=settings.authorization_url,
        scopes=_prepare_scopes(settings.scopes)
    )


async def verify_token(
    security_scopes: SecurityScopes,
    token: str = Security(get_oauth2_scheme)
) -> dict[str, Any]:
    # logic for verification of token and scopes
    # let's say that it returns decoded payload
    # in form of a dict
    ...
```
but (probably as expected) in this example `get_oauth2_scheme` is callable and what is actually passed to the function is `OAuth2AuthorizationCodeBearer` instance which hasn't been called to retrieve token from `Authorization` header. Also when getting to Swagger UI `Autorize` button is not rendered, but I guess this is more or less expected.

So I thought that perhaps putting another function in between that will use `Request` as dependency and `get_oauth2_scheme` as security dependency, then I will get callable `OAuth2AuthorizationCodeBearer` instance that I will call with `Request` instance and get a token (just as it is done underneath when used as `Dependcy` or `Security` argument).

```python
from typing import Any

from fastapi import Depends, Security
from fastapi.security import OAuth2AuthorizationCodeBearer, SecurityScopes

from our_project.settings import get_settings, Settings

def _prepare_scopes(scopes: str) -> dict[str, str]:
    ...


def get_oauth2_scheme(settings: Settings = Depends(get_settings)) -> OAuth2AuthorizationCodeBearer:
    return OAuth2AuthorizationCodeBearer(
        tokenUrl=settings.token_url,
        authorizationUrl=settings.authorization_url,
        scopes=_prepare_scopes(settings.scopes)
    )


async def get_token(
    request: Request,
    oauth2_scheme: OAuth2AuthorizationCodeBearer = Depends(get_oauth2_scheme)
) -> str:
    return await oauth2_scheme(request)


async def verify_token(
    security_scopes: SecurityScopes,
    token: str = Security(get_token)
) -> dict[str, Any]:
    # logic for verification of token and scopes
    # let's say that it returns decoded payload
    # in form of a dict
    ...
```
and here I get the token in my `verify_token` function, however `Authorize` button is not rendered in Swagger UI. I suspect that it is because `OAuth2AuthorizationCodeBearer` instance isn't passed as callable argument to `Security` and it can't read proper attributes that are declared in that instance and are necessary to render proper OpenAPI schema.

So I'm definitely missing some piece on how to make it working by passing dependencies. I'm clearly missing a piece where I can just obtain callable instance of `OAuth2AuthorizationCodeBearer` and pass it as argument to `Depends`/`Security`. Do you have some suggestion how to do this?

I know that two possible workarounds are to just invoke `get_settings` in global scope of our module and then instantiating `OAuth2AuthorizationCodeBearer` and keeping it in global variable which will then be used as a dependency in our function that verifies token and checks scopes:
```python
from typing import Any

from fastapi import Security
from fastapi.security import OAuth2AuthorizationCodeBearer, SecurityScopes

from our_project.settings import get_settings


def _prepare_scopes(scopes: str) -> dict[str, str]:
    ...


settings = get_settings()
oauth2_security_scheme = OAuth2AuthorizationCodeBearer(
    tokenUrl=settings.token_url,
    authorizationUrl=settings.authorization_url,
    scopes=_prepare_scopes(settings.scopes)
)


def verify_token(
    security_scopes: SecurityScopes,
    token: str = Security(oauth2_security_scheme)
) -> dict[str, Any]
    # logic for verification of token and scopes
    # let's say that it returns decoded payload
    # in form of a dict
    ...
```

Another method would be to enclose it in a function, calling `get_settings` there and return instance of `OAuth2AuthorizationCodeBearer`

```python
from typing import Any

from fastapi import Security
from fastapi.security import OAuth2AuthorizationCodeBearer

from our_project.settings import get_settings


def _prepare_scopes(scopes: str) -> dict[str, str]:
    ...


def get_oauth2_security_scheme() -> OAuth2AuthorizationCodeBearer:
    settings = get_settings()
    return OAuth2AuthorizationCodeBearer(
        tokenUrl=settings.token_url,
        authorizationUrl=settings.authorization_url,
        scopes=_prepare_scopes(settings.scopes)
    )


def verify_token(
    security_scopes: SecurityScopes,
    token: str = Security(get_oauth2_security_scheme())
) -> dict[str, Any]
    # logic for verification of token and scopes
    # let's say that it returns decoded payload
    # in form of a dict
    ...
```
but still, this second exmaple involves calling `get_oauth2_security_scheme` function which is for example called when we import this module elsewhere.

We wanted to avoid calling `get_settings` in global scope or calling `get_settings` when function definition is evaluated as this causes a yaml file (in which we store our configuration) to be read for example when this module is imported for testing. Also we think that FastAPI's dependency system is very convenient and eases up a lot, especially with override system when it comes to testing, that's why we'd like to use it in this case.

Perhaps solution with dependency injection is really easy here but I'm just missing something, so I'd really appreciate your help.

### Operating System

Linux

### FastAPI Version

0.70.1

### Python Version

3.9.5

**Comment by adriangb** [comment, marked as answer]
Likely a dupe of https://github.com/tiangolo/fastapi/issues/3317
